### PR TITLE
fix(compiler): improve the error when template is not a string

### LINKED ISSUE #8708: Improve the error message when a component's template is not a string
Today after some refactorings, I was faced with this stack trace upon launching my application:

```
Error: Uncaught (in promise): TypeError: Object doesn't support property or method 'charCodeAt'
   at resolvePromise (http://localhost:4146/vendor-dev.js:99618:27)
   at Anonymous function (http://localhost:4146/vendor-dev.js:99595:15)
   at Anonymous function (http://localhost:4146/vendor-dev.js:7373:26)
   at ZoneDelegate.prototype.invoke (http://localhost:4146/vendor-dev.js:99401:15)
   at onInvoke (http://localhost:4146/vendor-dev.js:9189:26)
   at ZoneDelegate.prototype.invoke (http://localhost:4146/vendor-dev.js:99401:15)
   at Zone.prototype.run (http://localhost:4146/vendor-dev.js:99296:19)
   at Anonymous function (http://localhost:4146/vendor-dev.js:99651:19)
   at ZoneDelegate.prototype.invokeTask (http://localhost:4146/vendor-dev.js:99434:19)
   at onInvokeTask (http://localhost:4146/vendor-dev.js:9180:26)
vendor-dev.js (36611,14)
```

The reason for this stack trace was that I had omitted a `.html` when `require`ing a component's template. This resulted in webpack adding a default `.ts` extension and using my component's module instead of the actual HTML template and the component itself became the template.

The same effect and stack trace can be achieved by

``` ts
@Component({
    //....
    template: <string>{}
})
```

I think there should be a better error message / stack trace to quicker get to the bottom of this.

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Closes #8708

### CODE REVIEW

File: modules/@angular/compiler/test/directive_normalizer_spec.ts [hunk 1]
```diff
@@ -33,6 +33,22 @@ export function main() {
...
          }));
+      it('should throw if template is not a string',
+         inject([DirectiveNormalizer], (normalizer: DirectiveNormalizer) => {
+           expect(() => normalizer.normalizeTemplate({
+             componentType: SomeComp,
+             moduleUrl: SOME_MODULE_URL,
+             template: <any>{}
+           })).toThrowError('The template specified for component SomeComp is not a string');
+         }));
+      it('should throw if template url is not html file',
```
Comment (vicb) [inline modules/@angular/compiler/test/directive_normalizer_spec.ts]: I'n not sure if we need this test ? Seems too restrictive.

Could you explain the rationale ?

Comment (vicb) [inline modules/@angular/compiler/test/directive_normalizer_spec.ts]: yep similar:
- people can use whatever extension they want,
- an `.html` extension is not a guarantee that the file is valid html.

**Comment by DzmitryShylovich** [comment]
@vicb comment is addressed

File: modules/@angular/compiler/src/directive_normalizer.ts [hunk 2]
```diff
@@ -64,10 +64,18 @@ export class DirectiveNormalizer {
       SyncAsyncResult<CompileTemplateMetadata> {
     let normalizedTemplateSync: CompileTemplateMetadata = null;
     let normalizedTemplateAsync: Promise<CompileTemplateMetadata>;
-    if (prenormData.template) {
+    if (prenormData.template != null) {
```
Comment (vicb) [inline modules/@angular/compiler/src/directive_normalizer.ts]: `if (.. && ..)`

**Comment by vicb** [comment]
Could you please rebase ?

**Comment by DzmitryShylovich** [comment]
@vicb done

**Comment by vicb** [comment]
Thanks !

**Pull request closed without merging**